java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -312,10 +344,36 @@ abstract class AbstractArcHost(
      *
      * Subclasses may override this to store the [context] using a different implementation.
      */
-    protected open suspend fun writeContextToStorage(arcId: String, context: ArcHostContext) =
-        createArcHostContextParticle(context).run {
-            writeArcHostContext(context.arcId, context)
+    protected open suspend fun writeContextToStorage(arcId: String, context: ArcHostContext) {
+        /** Serialize the [context] to storage in observed order asynchronously. */
+        if (!channel.isClosedForSend) {
+            var contextCopy: ArcHostContext? = ArcHostContext(
+                context.arcId,
+                context.particles,
+                context.handleManager,
+                context.arcState
+            )
+            channel.send {
+                try {
+                    contextCopy?.let {
+                        /** TODO: reuse [ArcHostContextParticle] instances if possible. */
+                        createArcHostContextParticle(it).run {
+                            writeArcHostContext(it.arcId, it)
+                        }
+                        // Remove the reference in the flow/channel cache.
+                        contextCopy = null
```

Comment:
Why is this needed? Where is this being cached?

Reply:
Please see:
https://github.com/PolymerLabs/arcs/pull/5352
https://github.com/PolymerLabs/arcs/pull/5413

The reference is held by `result in FlowKt_ChannelsKt...`

Reply:
This seems like another argument against flows. Will this issue occur if we're just using channels?

Reply:
It only occurs in flows. iirc at flow's `emitAll` function.